### DISCUSSION TITLE
Css for non-rendered components gets added to the bundle

### DISCUSSION THREAD

**Discussion by Niqp** [vercel/next.js#69594] [Help]
### Summary

Hi everyone! We're currently trying to move our project to NextJs with an App directory.
We're using dynamic routing, because our site is highly modular, rendering different components in a single page based on the data from db.

We are styling each component by importing SCSS modules, and in parent components, we conditionally choose whether to render a component or not based on the data we got from the request.  All of the components in question are rendered server-side and we would like to keep it that way.

The issue is, the css styles from components get added to the client bundle, even if they are not actually rendered. As soon as we import the component at the top, we can locate the CSS classes when inspecting the page, although the component itself is not in the markup.

(Example JSX:
`{!!condition ? <Component1> : <Component2>}`
CSS for both components gets sent to the client even if it wasn't rendered on the server at all)

Our site is highly performance-dependent, so we would like to send as little of unused CSS to the client as possible.

What we tried:
1. using dynamic() or lazy() on the component - all of those only work on client components, so applying those had no effect.
2. Suspense -  all our data is fetched together so its not helpful
3. use CSS Hooks instead of modules - this has another issue for us, because all the components are added to the __next.push() script part of the page (for hydration, as I understand), it means that all the inline styles would get copied there as well, as many times as we are using the component, which bloats the html size significantly. 
4. CSS-In-JS is currently also only supported for client components.

Can anyone please suggest a way for us to only send to the client the CSS of the components that were actually rendered on the server?

**Comment by danielleHerszfang** [comment]
Also having this issue!

**Comment by FranCarstens** [comment]
Have you found a solution?

As far as I've been able to find out, this is a known issue; we have the same problem with a platform designed in a similar fashion. The dynamic nature of your route prevents tree shaking unused css because Webpack can't determine which CSS will in fact be used. You can see even Next's [experimental solution to unused CSS](https://nextjs.org/docs/app/api-reference/config/next-config-js/cssChunking) specifically states: "This allows you to load only the CSS that is needed for a **specific route**, instead of loading all the application's CSS at once."

If that **specific route** is dynamic it will still include all the possible css, even if it's never actually used.

It looks like the best move is to switch to a utility first CSS framework like Tailwind, sorry :(. It’s a big lift, and your global CSS will grow, but you'll be able to get rid of a large chunk of your CSS modules.
